### PR TITLE
[DeviceMesh] Add a private _flatten() API for device_mesh

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132779
* __->__ #132632

Adds a new private API to flatten a DeviceMesh to a 1D DeviceMesh such that:
```
mesh_3d = init_device_mesh(
    self.device_type, (2, 2, 2), mesh_dim_names=("dp", "cp", "tp"),
)

dp_cp_mesh = mesh_3d["dp", "cp"]
# flattened_mesh on rank 0, 2, 4, 6 is DeviceMesh([0, 2, 4, 6], mesh_dim_names=('dp_cp',))
# flattened_mesh on rank 1, 3, 5, 7 is DeviceMesh([1, 3, 5, 7], mesh_dim_names=('dp_cp',))
flattened_dp_cp_mesh = dp_cp_mesh._flatten()
```

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wconstab @d4l3k @c-p-i-o @tianyu-l

### CODE REVIEW

File: torch/distributed/device_mesh.py
Comment (wz337) [inline torch/distributed/device_mesh.py:116]: This is similar to `create_sub_mesh`. The main difference is we need to sort the `mesh_dim_names` to flatten, get the mesh size of the flatten dim, and _init_backend needs to be True. 

We can probably see if we can merge this into `create_sub_mesh` later but want to have this out so we can't unblock the dp cp use case.

File: torch/distributed/device_mesh.py
Comment (wanchaol) [inline torch/distributed/device_mesh.py:693]: why can't we make this a public API?

File: torch/distributed/device_mesh.py
Comment (wanchaol) [inline torch/distributed/device_mesh.py:702]: nit: `return _mesh_resources.create_flatten_mesh(self)`

File: torch/distributed/device_mesh.py
Comment (wanchaol) [inline torch/distributed/device_mesh.py:695]:
```suggestion
            Returns a 1D DeviceMesh by flattening the current DeviceMesh.
```

File: test/distributed/test_device_mesh.py
Comment (wanchaol) [inline test/distributed/test_device_mesh.py:586]: Is there a reason for this ordered behavior? I feel if user want to swap the mesh dim orders, and flatten the swapped order device mesh, we should probably respect that?

**Review by wanchaol — COMMENTED** [review]
changes looks good! Have some comments/questions inlined

File: torch/distributed/device_mesh.py
Comment (wz337) [reply torch/distributed/device_mesh.py:693]: I was thinking let Chien-chin try it out first to see if there is any minor UX improvement we want to make before making it public. Wdyt?

File: test/distributed/test_device_mesh.py
Comment (wz337) [reply test/distributed/test_device_mesh.py:586]: Say if we have a 3D mesh and we create mesh["dp", "cp"] and mesh["cp", "dp"], the result does look different and the order matters, because of the 2D mesh_dim order. However, if we flatten out these two, the result would both be a 1D mesh with the same global ranks, except that the mesh tensor's order is different and the mesh_dim_names are going to be different.
```
DeviceMesh([0, 4, 1, 5], mesh_dim_names=('dp_cp',))
DeviceMesh([0, 1, 4, 5], mesh_dim_names=('cp_dp',))
```
If we distinguish these two, then we will create two pgs with different local rank result. I am not sure if we want to distinguish these two.

File: torch/distributed/device_mesh.py
Comment (fegin) [reply torch/distributed/device_mesh.py:116]: Do we want to let user to pass the name as we discussed before? Sometimes reusing the existing dim names can be confusing.

File: torch/distributed/device_mesh.py
Comment (fegin) [inline torch/distributed/device_mesh.py:702]: Are we going to cache the flattened mesh or this will be done in the following PR?

**Review by fegin — COMMENTED** [review]
Nice, let me try with FSDP + CP.

update: verified with DP2CP4 and DP4CP2. Both run without issues.

File: torch/distributed/device_mesh.py
Comment (wz337) [reply torch/distributed/device_mesh.py:702]: Will include it in the next PR.

**Review by fegin — APPROVED** [review]
LGTM, the PR works with FSDP + CP. The API improvement can be done in separate PRs.

File: test/distributed/test_device_mesh.py
Comment (wanchaol) [reply test/distributed/test_device_mesh.py:586]: I think this would bring some user confusions about the mesh dim order, If we don't want to distinguish these two, we should simply not allowing reordered mesh dim slicing (i.e. we only allow submesh slicing happen in a contiguous manner like before, but it could still skip the dims in the middle)

File: torch/distributed/device_mesh.py
Comment (wanchaol) [reply torch/distributed/device_mesh.py:696]: sure we can make it public later, I don't think it's quite a concern though given that we are working on the main branch which is separate from the release schedule, but your call

File: test/distributed/test_device_mesh.py
Comment (wz337) [reply test/distributed/test_device_mesh.py]: Sounds good. I removed the sorting part and the test for flattening with swapping dims in this PR. 
I will put up a separate PR to add the restriction for slicing.

**Review by wanchaol — APPROVED** [review]
lgtm

**Pull request closed without merging**
